internal/codeintel/bundles/persistence/cache/reader_cache.go

```diff
@@ -0,0 +1,228 @@
+			// Decrement refcount
+			if atomic.AddUint32(&entry.refCount, ^uint32(0)) == 0 {
+				// Wake routines waiting for this value to go to zero
+				entry.refCountChanged.Broadcast()
+
+				// No more readers, update expiry
+				entry.expiry = time.Now().UTC().Add(c.ttl)
+			}
+		}()
+
+		select {
+		case <-entry.init:
+			if entry.err != nil {
+				return entry.err
+			}
+			return fn(entry.reader)
+
+		case <-ctx.Done():
+		}
+	}
+
+	return ErrReaderInitializationDeadlineExceeded
+}
+
+// getOrCreateActiveEntry gets or creates a cache entry for the given key. If there exists an entry for
+// the key, but that entry is marked as draining, we wait until that entry has completely drained and then
+// try again.
+func (c *Cache) getOrCreateActiveEntry(ctx context.Context, key string) (*cacheEntry, bool) {
+loop:
+	for {
+		entry, ok := c.getOrCreateRawEntry(key)
+		if ok {
+			return entry, true
+		}
+
+		select {
+		case <-entry.drained:
+			continue loop
+		case <-ctx.Done():
+			return nil, false
+		}
+	}
+}
+
+// getOrCreateRawEntry gets or creates a cache entry for the given key and a boolean flag indicating
+// whether or not the entry is "active". True indicates that the entry is not currently marked as
+// draining and can be used. In this case the entry's wait group value is not modified and the entry's
+// close procedure is unaffected.
+func (c *Cache) getOrCreateRawEntry(key string) (*cacheEntry, bool) {
+	c.m.Lock()
+	defer c.m.Unlock()
+
+	entry, ok := c.entries[key]
+	if !ok {
+		entry = c.makeEntry(key)
+		c.entries[key] = entry
+	}
+
+	if !entry.draining {
+		// Mark as in-use and disable expiry while held
+		atomic.AddUint32(&entry.refCount, 1)
+		entry.expiry = time.Time{}
+	}
+
+	return entry, !entry.draining
+}
+
+// makeEntry creates a new empty cache entry. This will be
+func (c *Cache) makeEntry(key string) *cacheEntry {
+	entry := &cacheEntry{
+		init:            make(chan struct{}),
+		drained:         make(chan struct{}),
+		refCountChanged: sync.NewCond(&sync.Mutex{}),
+	}
+
+	go func() {
+		// Mark as initialized
+		defer close(entry.init)
+
+		if entry.reader, entry.err = c.opener(key); entry.err != nil {
+			// Do not hold on to cache entries that failed to initialize. In the case of a
+			// transient error (e.g., a missing bundle file that is later uploaded), we do
+			// not want a poison cache value occupying that key until a process restart.
+			c.m.Lock()
+			delete(c.entries, key)
+			c.m.Unlock()
+		}
+	}()
+
+	return entry
+}
+
+// evict iterates through all entries and determines those which have expired. These entries are marked
+// for removal which will occur once initialization finishes and any active readers have detached. Generally,
+// no readers should be active, but may occur when a request around the expiry races.
+func (c *Cache) evict(now time.Time) {
+	c.m.Lock()
+	defer c.m.Unlock()
+
+	for key, entry := range c.entries {
+		if !entry.expiry.IsZero() && entry.expiry.Before(now) {
```

Comment:
nit: strictly, this should be 
```suggestion
		if !entry.expiry.IsZero() && !entry.expiry.After(now) {
```

assuming right at `expiry` also means expired?

Reply:
tbh I think the code reads nicer with the use of Before. We don't need to be strict on the interpretation of expire.